wallet/src/key_chain.rs

```diff
@@ -17,184 +17,619 @@
+    ReceiveFunds,
     /// This is for the internal usage of the wallet when creating change output for a transaction
-    Change = 1,
+    #[codec(index = 1)]
+    Change,
+}
+
+impl KeyPurpose {
+    const ALL: [KeyPurpose; 2] = [KeyPurpose::ReceiveFunds, KeyPurpose::Change];
+    /// The index for each purpose
+    const DETERMINISTIC_INDEX: [ChildNumber; 2] = [
+        ChildNumber::from_normal(U31::from_u32_with_msb(0).0),
+        ChildNumber::from_normal(U31::from_u32_with_msb(1).0),
+    ];
+}
+
+impl TryFrom<ChildNumber> for KeyPurpose {
+    type Error = KeyChainError;
+
+    fn try_from(num: ChildNumber) -> Result<Self, Self::Error> {
+        match num.get_index() {
+            0 => Ok(ReceiveFunds),
+            1 => Ok(Change),
+            _ => Err(KeyChainError::InvalidKeyPurpose(num)),
+        }
+    }
 }
 
 #[allow(dead_code)] // TODO remove
-/// This key chain contains a pool of pre-generated keys and addresses for the usage in a wallet
-pub struct KeyChain<B: Backend> {
+pub struct MasterKeyChain {
+    // pub struct MasterKeyChain<B: Backend> {
     /// The specific chain this KeyChain is based on, this will affect the address format
     chain_config: Arc<ChainConfig>,
-
     /// The master key of this key chain from where all the keys are derived from
     // TODO implement encryption
-    master_key: ExtendedPrivateKey,
-
-    /// The database connection
-    // TODO Consider if this should be an Option
-    db: Arc<Store<B>>,
-
-    /// The number of addresses to pre-generate from the master key
-    lookahead_size: u16,
+    root_key: ExtendedPrivateKey,
 }
-#[allow(dead_code)] // TODO remove
-impl<B: Backend> KeyChain<B> {
-    pub fn new_from_mnemonic(
+
+impl MasterKeyChain {
+    #[allow(dead_code)] // TODO remove
+    pub fn new_from_mnemonic<B: Backend>(
         chain_config: Arc<ChainConfig>,
+        db_tx: &mut StoreTxRw<B>,
         mnemonic_str: &str,
         passphrase: Option<&str>,
-        db: Arc<Store<B>>,
     ) -> KeyChainResult<Self> {
-        // TODO use Mnemonic::parse when bip39 is configured with "std"
-        let mnemonic =
-            bip39::Mnemonic::parse_normalized(mnemonic_str).map_err(KeyChainError::Bip39)?;
-        // TODO use Mnemonic::to_seed when bip39 is configured with "std"
-        let mut seed = mnemonic.to_seed_normalized(passphrase.unwrap_or(""));
-        let master_key = ExtendedPrivateKey::new_master(&seed, KEY_KIND)?;
-        // TODO(SECURITY) erase mnemonic
+        let mut mnemonic = bip39::Mnemonic::parse(mnemonic_str).map_err(KeyChainError::Bip39)?;
+        let mut seed = mnemonic.to_seed(passphrase.unwrap_or(""));
+        let root_key = ExtendedPrivateKey::new_master(&seed, DEFAULT_KEY_KIND)?;
+        mnemonic.zeroize();
         seed.zeroize();
 
-        Ok(KeyChain {
+        let key_id = KeyId::DeterministicRoot(root_key.to_public_key());
+        let key_content = KeyContent::DeterministicRoot(root_key);
+
+        db_tx.set_key(&key_id, &key_content)?;
```

Comment:
It would probably be cleaner to separate the functionality of deriving the root key from the seed and storing it in the database.